src/pyggdrasil/scripts/cell_simulation.py

```diff
@@ -0,0 +1,369 @@
+    pass
+
+
+def gen_sim_data(
+    params: dict,
+    rng: PRNGKeyArray,
+    *,
+    tree_no: int,
+) -> None:
+    """
+    Generates cell mutation matrix for one tree and writes to file.
+
+    Args:
+        params: dict
+            input parameters from parser for simulation
+        rng: JAX random number generator
+        tree_no: int - optional
+            tree number if a series is generated
+    Returns:
+        None
+    """
+    ############################################################################
+    # Parameters
+    ############################################################################
+    outdir = params["outdir"]
+    n_cells = params["n_cells"]
+    n_mutations = params["n_mutations"]
+    alpha = params["alpha"]
+    beta = params["beta"]
+    na_rate = params["na_rate"]
+    observe_homozygous = params["observe_homozygous"]
+    strategy = params["strategy"]
+    verbose = params["verbose"]
+
+    ############################################################################
+    # Random Seeds
+    ############################################################################
+    rng_tree, rng_cell_attachment, rng_noise = random.split(rng, 3)
+
+    ##############################################################################
+    # Generate Trees
+    ##############################################################################
+    #  generate random trees (uniform sampling) as adjacency matrix
+    tree_inv = generate_random_tree(rng_tree, n_nodes=n_mutations)
+    # reverse node order
+    tree = reverse_node_order(tree_inv)
+
+    ##############################################################################
+    # Attach Cells To Tree
+    ###############################################################################
+    # convert adjacency matrix to self-connected tree - in tree_inference format
+    np.fill_diagonal(tree, 1)
+    # define strategy
+    strategy = sim.CellAttachmentStrategy[strategy]
+    # attach cells to tree - generate perfect mutation matrix
+    perfect_mutation_mat = sim.attach_cells_to_tree(
+        rng_cell_attachment, tree, n_cells, strategy
+    )
+
+    ###############################################################################
+    # Add Noise
+    ################################################################################
+    # add noise to perfect mutation matrix
+    noisy_mutation_mat = None
+    if (beta > 0) or (alpha > 0) or (na_rate > 0):
+        noisy_mutation_mat = sim.add_noise_to_perfect_matrix(
+            rng_noise, perfect_mutation_mat, alpha, beta, na_rate, observe_homozygous
+        )
+
+    ################################################################################
+    # Save Simulation Results
+    ################################################################################
+    # make save name and path from parameters
+    filename = compose_save_name(params, tree_no=tree_no) + ".json"
```

Comment:
(Educational comment: feel free to resolve.)

`os` and `os.path` of course work and are a traditional way of working with the filesystem, but Python 3 has a nice new [pathlib library](https://docs.python.org/3/library/pathlib.html) with `pathlib.Path` object. I'm pretty sure you'll like it :wink:

Reply:
Will have a look for next time.